prost/src/encoding.rs

```diff
@@ -297,9 +297,14 @@ macro_rules! varint {
 
             #[inline]
             pub fn encoded_len_repeated(tag: u32, values: &[$ty]) -> usize {
-                key_len(tag) * values.len() + values.iter().map(|$to_uint64_value| {
-                    encoded_len_varint($to_uint64)
-                }).sum::<usize>()
+                let len = key_len(tag) as u64 * values.len() as u64
+                    + values
+                        .iter()
+                        .map(|$to_uint64_value| {
+                            encoded_len_varint($to_uint64) as u64
+                        })
+                        .sum::<u64>();
+                len.try_into().expect("encoded length overflows usize")
```

Comment:
> I am not convinced that panic improves the situation.

This is better than silently returning an overflowed length, which is the status quo with release builds.

> Should encoded_len return a Result<u64, MessageTooLargeError>?

Yes, but I meant this PR to be merged for a 0.13.x release so that current applications could be patched without API breaks.

In a later change, we can change the signature of `encoded_len` and grep for all occurrences of `.expect("encoded length overflows usize")` (I took care to use this distinctive panic message everywhere) and convert them to error returns.